libraries/chain/wasm_interface_collection.cpp

```diff
@@ -0,0 +1,128 @@
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (eosvmoc_tierup != wasm_interface::vm_oc_enable::oc_none) {
+      EOS_ASSERT(vm != wasm_interface::vm_type::eos_vm_oc, wasm_exception, "You can't use EOS VM OC as the base runtime when tier up is activated");
+      eosvmoc = std::make_unique<eosvmoc_tier>(data_dir, eosvmoc_config, d);
+   }
+#endif
+}
+
+wasm_interface_collection::~wasm_interface_collection() = default;
+
+void wasm_interface_collection::apply(const digest_type& code_hash, const uint8_t& vm_type, const uint8_t& vm_version, apply_context& context) {
+   if (substitute_apply && substitute_apply(code_hash, vm_type, vm_version, context))
+      return;
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (eosvmoc && (eosvmoc_tierup == wasm_interface::vm_oc_enable::oc_all || context.should_use_eos_vm_oc())) {
+      const chain::eosvmoc::code_descriptor* cd = nullptr;
+      chain::eosvmoc::code_cache_base::get_cd_failure failure = chain::eosvmoc::code_cache_base::get_cd_failure::temporary;
+      try {
+         const bool high_priority = context.get_receiver().prefix() == chain::config::system_account_name;
+         cd = eosvmoc->cc.get_descriptor_for_code(high_priority, code_hash, vm_version, context.control.is_write_window(), failure);
+      } catch (...) {
+         // swallow errors here, if EOS VM OC has gone in to the weeds we shouldn't bail: continue to try and run baseline
+         // In the future, consider moving bits of EOS VM that can fire exceptions and such out of this call path
+         static bool once_is_enough;
+         if (!once_is_enough)
+            elog("EOS VM OC has encountered an unexpected failure");
+         once_is_enough = true;
+      }
+      if (cd) {
+         if (!context.is_applying_block()) // read_only_trx_test.py looks for this log statement
+            tlog("${a} speculatively executing ${h} with eos vm oc", ("a", context.get_receiver())("h", code_hash));
+         eosvmoc->exec->execute(*cd, eosvmoc->mem, context);
+         return;
+      }
+   }
+#endif
+   if (is_on_main_thread()
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+       || is_eos_vm_oc_enabled()
+#endif
+   ) {
+      return wasmif.apply(code_hash, vm_type, vm_version, context);
+   }
+   threaded_wasmifs[std::this_thread::get_id()]->apply(code_hash, vm_type, vm_version, context);
+}
+
+// update current lib of all wasm interfaces
+void wasm_interface_collection::current_lib(const uint32_t lib) {
+   // producer_plugin has already asserted irreversible_block signal is called in write window
+   std::function<void(const digest_type& code_hash, uint8_t vm_version)> cb{};
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (eosvmoc) {
+      cb = [&](const digest_type& code_hash, uint8_t vm_version) {
+         eosvmoc->cc.free_code(code_hash, vm_version);
+      };
+   }
+#endif
+   wasmif.current_lib(lib, cb);
+   for (auto& w : threaded_wasmifs) {
+      w.second->current_lib(lib, cb);
```

Comment:
(I haven't studied this PR closely yet so apologies if this should be obvious..)

This part feels a little strange to me: there should only be a single OC code cache that is shared across all threads. So why pass the `cb` here that ` eosvmoc->cc.free_code(code_hash, vm_version)` per thread (presumably it gets called more than once).

Reply:
Each `wasm_interface` has its own `wasm_instantiation_cache`. There is only one `eosvmoc`. This makes sure that all the codes are freed no matter which `wasm_interface`s caches it.